### DISCUSSION TITLE
ForEachReducer support for dictionary-based collections

### DISCUSSION THREAD

**Discussion by heyltsjay** [pointfreeco/swift-composable-architecture#1365] [Beta]
`_ForEachReducer` drops support for dictionary-based collections, which was unexpected and not documented in the migration guide. Should either support or call out the design decision to move away from it. 

Our app (UIKit) currently models forms using dictionaries, mostly to help coalesce actions that come in:

```
struct State {
  var form: [FormFieldID: FormField.State] = [
     .email: .init(),
     .password: .init(),
  ]
}

enum Action {
  case form(id: FormFieldID, action: FormField.Action)
}
```

**Comment by heyltsjay** [comment]
If it's just an oversight, migrators can use this for now:

```swift
    @inlinable
    func forEach<Element: ReducerProtocol, Key>(
      _ toDictionaryState: WritableKeyPath<State, [Key: Element.State]>,
      action toKeyedAction: CasePath<Action, (Key, Element.Action)>,
      @ReducerBuilderOf<Element> _ element: @escaping () -> Element,
      file: StaticString = #file,
      fileID: StaticString = #fileID,
      line: UInt = #line
    ) -> CombineReducers<ReducerBuilder<Self.State, Self.Action>._Sequence<Reduce<Self.State, Self.Action>, Self>> {
        CombineReducers {
            Reduce { state, action in
                AnyReducer {
                    element()
                }
                .forEach(
                    state: toDictionaryState,
                    action: toKeyedAction,
                    environment: { _ in },
                    file: file,
                    fileID: fileID,
                    line: line
                )
                .callAsFunction(&state, action, ())
            }
            self
        }
    }
```

**Comment by stephencelis** [comment]
Thanks for bringing this up @heyltsjay! I think like the other operators, we want to take some time to re-envision the use of this one, which we haven't used ourselves in practice.

The main a few things that come to mind:

- We should flip it to be defined in the new reducer builder way (as you do in your migration sketch, above).
- I think we should probably consider requiring `OrderedDictionary`. Right now if you try to render a `Dictionary` to SwiftUI view state, you need to somehow loop over the dictionary to render it, you need to order the keys in some way to avoid getting randomness on recomputation.
- I wonder if we should also have a `ForEachStore` that is compatible with `OrderedDictionary`.

Does anyone out there have some real world experience using `Reducer.forEach` with dictionary state? Would love for some examples (and maybe a case study) to work with to help drive the design here.

**Reply by jshier** [reply]
> * I think we should probably consider requiring `OrderedDictionary`. Right now if you try to render a `Dictionary` to SwiftUI view state, you need to somehow loop over the dictionary to render it, you need to order the keys in some way to avoid getting randomness on recomputation.

This isn't strictly true. While `Dictionary`s don't have constant ordering, the order is static for each run of the app. So looping over the same value multiple times will have the same ordering. Of course mutations will be arbitrarily inserted, so if that's a concern you'd still want an `OrderedDictionary`.

**Reply by stephencelis** [reply]
> Of course mutations will be arbitrarily inserted, so if that's a concern you'd still want an `OrderedDictionary`.

That's the concern, yup! Considering TCA state is generally mapped to UIs, we should be using data structures with predictable ordering.

**Reply by tgrapperon** [reply]
I think I'm using some in my app where I found `IdentifiedArray` too cumbersome to work with.
I a few words, I'm using a dictionary in adjunction to a regular `Array` (or `IdentifiedArray`) to form a kind of unstructured `IdentifiedArray`.

The dictionary hosts the whole set of child states that I don't want to discard, and that I don't want to bother ordering. The array hosts a temporary slice of these states as I'm filtering during searches (think the content of the cells in a list). If the cells are simple, I'm using a standard `ForEach` over some `ViewStore's array of features. If the cells are complex, I'm using a `ForEachStore` where each cell is a feature itself.

The content of these cells during search is dependent from the search terms, and is not part of the model of the more complex features hosted in the dictionary and from where you can navigate to.

When the user searches, I'm updating the `(Identified)Array` of cells to represent the searched list. When the user navigates, I'm using a `.forEach` reducer over the dictionary of states to back the complete feature.

I could probably model everything with `IdentifiedArray`, but it would be much more cumbersome. I later formalized a little more this approach with my "lazy foreach" experiment, because I needed to upgrade the dictionary states on the fly too. I didn't pushed farther yet to improve the ergonomics, as I feel it's only one particular representation of a general derived state thing I'm working on from time to time.

So to sum up, I think they can be useful when you want to extract a disjoint subset for presentation. Of course an `IdentifiedArray` can do the same, but it brings an ordering dimension that is not always meaningful for the feature. In my case, the cell's feature is not a sub-feature of the child feature because it mixes search terms to highlight search results, so I opted to keep them disjoints. Of course I can model the thing differently, but I think it will complexify even more something that is already quite complex.

**Reply by stephencelis** [reply]
> The dictionary hosts the whole set of child states that I don't want to discard, and that I don't want to bother ordering.

Would you be able to use `OrderedDictionary` and ignore the ordering if that's the interface we decided to ship?

**Reply by iampatbrown** [reply]
I've found it useful to generalise things a bit and use:

```swift
toIds: KeyPath<State, OrderedSet<ID>>,
toElementState: @escaping (ID) -> WritableKeyPath<State, Element.State?>,
```

then for `IdentifiedArray` I'd use:
```swift
toIds: toElementsState.appending(path: \.ids),
toElementState: { id in toElementsState.appending(path: \.[id: id]) },
```
and `OrderedDictionary` would be:
```swift
toIds: toElementsState.appending(path: \.keys),
toElementState: { key in toElementsState.appending(path: \.[key]) },
```

I'd mainly be using it for computed features though and add a subscript for creating the feature with shared state. 

Here's a branch demonstrating what I mean. https://github.com/pointfreeco/swift-composable-architecture/compare/protocol-beta...iampatbrown:swift-composable-architecture:for-each-ids

I haven't tested it at all, it's just to give a better idea of what I'm talking about.

**Reply by tgrapperon** [reply]
@iampatbrown The computed feature thing and the abstracted/extracted `IDs` reminds me a little #1227. I didn't look in detail at your implementation yet, but it is possible that they could overlap/generalize.

**Reply by iampatbrown** [reply]
@tgrapperon yeah it does look like there's some overlap. I've played with something similar before around the idea of `ComputedState. Ultimately, I dropped back to using subscripts because it required fewer changes and was easier for me to implement across projects.

I haven't explored it in a while though and #1227 looks pretty interesting :)

**Reply by tgrapperon** [reply]
@stephencelis Is there something that prevents from publicizing the internal `OrderedDictionary` of `IdentifiedArray`, in the form of:
```swift
extension OrderedDictionary {
  public init(_ identifiedArray: IdentifiedArray<Key, Value>) {
    self = identifiedArray._dictionary
  }
}
```
This operation is very efficient and one could then base everything on `OrderedDictionary` and provide specializations for `IdentifiedArray`, as the `Identifiable` nature doesn't play a role in `Reducer.ForEach`. I looked at the `swift-identified-collections`, but I didn't find anything like that.

In this case, a `ForEachStore` for `OrderedDictionary` would make sense indeed, using the same technique.

**Reply by tgrapperon** [reply]
[Here is](https://github.com/pointfreeco/swift-composable-architecture/compare/protocol-beta...tgrapperon:swift-composable-architecture:protocol-beta-generalized-foreach?expand=1) a quick experiment with completely generic `Reducer.ForEach` and `ForEachStore`. It doesn't seems to cause additional strain on compiler.

As I'm getting it, any type conforming to the `IdentifiedStatesCollection` protocol could work. I've provided conformance for `IdentifiedArray` and `OrderedDictionary`, and it should also work with unstructured or lazier collections. This is a very fast proof of concept, and this protocol could likely be refined.

What do you think?

**Reply by tgrapperon** [reply]
As a byproduct of the generalization, one can `ForEach` reducers and stores using different `IdentifiedStatesCollection` on the same set of data. It means that one can non destructively and efficiently filter or sort for the view layer while preserving the existing states for example.

**Reply by iampatbrown** [reply]
@tgrapperon That looks like it could work. I'd probably require `IDs` be an `OrderedSet` and remove `areIdentifiersEqual`.

I'm hesitant about introducing a public protocol for this though. While the flexibility of using custom collections with `ForEachStore` is appealing, I'm concerned it would add constraints to the  implementation. 

I think adding support for `OrderedDictionary` would cover the majority of use cases.

**Reply by iampatbrown** [reply]
Something like this:

```swift
public protocol IdentifiedCollection {
  associatedtype ID: Hashable
  associatedtype Value

  var ids: OrderedSet<ID> { get }
  subscript(id id: ID) -> Value? { get set }
}

extension IdentifiedArray: IdentifiedCollection {
  public typealias Value = Element
}

extension OrderedDictionary: IdentifiedCollection {
  public typealias ID = Key

  public var ids: OrderedSet<ID> { keys }
  public subscript(id id: ID) -> Value? {
    _read { yield self[id] }
    _modify { yield &self[id] }
  }
}
```

Then in `ForEachStore` you might be able to use:

```swift
 let elements = store.state.value
 self.data = elements.ids.lazy.map { elements[id: $0] }
```

Does `self.data` actually get used much?

**Reply by tgrapperon** [reply]
>Does self.data actually get used much?

It seems never! At least the one from `ForEachStore`. I've tried to delete/move/insert and it is never called. I even tried to `fatalError()` in a computed property, and it never touches it.

Furthermore, it seems there is an issue with the current implementation, as we capture `store.state.value` in `init`. If the `store` instance doesn't change but the number of views or content changes without invalidating the parent of `ForEachStore`, this value won't be updated.

It doesn't seem it affects edit operations.

I've managed to make `TODO` crash with a drop, but I suspect it is a vanilla problem. I'll try to reproduce the issue.

At worse, we can promote the `ViewStore` to `@ObservedObject` and decide that `Data == IDs`. This is not optimal, but it will already be better than the uncontrolled collection that is returned today. Given that `content` is escaping, if `ForEachStore` is evaluated, it will trickle down to `ForEach` at least, so it shouldn't impact performance. There is nothing in the doc that describes what this type represents, so it should probably work as fine with `IDs` that we're already keeping up-to-date anyway.

About the protocol thing, you're probably right. I'm still pushing with my `LazyForEach` to check if looser constraints are required, but I think it could work even with `OrderedSet` being fixed. I didn't use the name `IdentifiedCollection` as I thought it would collide with `swift-identified-collections` (and I don't think this protocol has its place there neither). 

I don't think one can easily make the protocol internal, and there are only 2 conforming types, so maybe the safest would be to to parametrize with functions like you suggested. It that can be upgraded later to a protocol if required.

**Comment by iampatbrown** [comment]
> I've managed to make TODO crash with a drop, but I suspect it is a vanilla problem. I'll try to reproduce the issue.

Interesting... let me know if you figure out how to reproduce it.

> At worse, we can promote the ViewStore to @ObservedObject and decide that Data == IDs

Could probably also do something like:

```swift
public var data: Data { state.value }
let state: CurrentValueSubject<Data, Never>
```

> I'm still pushing with my LazyForEach  

Which is understandable :) Looks like it would be a powerful tool. Maybe `LazyForEachStore` could be an option.

> I didn't use the name IdentifiedCollection as I thought it would collide with swift-identified-collections (and I don't think this protocol has its place there neither).

Yeah it isn't quite an `IdentifiedCollection` however I'd be leaning more towards a protocol in `swift-identified-collections` than here.

> maybe the safest would be to to parametrize with functions like you suggested. It that can be upgraded later to a protocol if required.

Yeah that makes sense.

**Reply by tgrapperon** [reply]
> Interesting... let me know if you figure out how to reproduce it.

Yes, this is Vanilla and it happens when you insert something in a `List` that comes from another section from the `List` (not a `move` intra-section). Not a TCA problem, and not reproducible with the current `Todo` without adding some code.

About `DynamicViewContent`, after having played a little with it, I don't think this is a thing for us to implement. The best we should do is pass wrapped ones up. It seems that `SwiftUI` only looks at `ForEach` anyway, which is understandable from their standpoint.

About my `LazyForEach`, I think that an `Update` Reducer could maybe be defined to update the `EachState` before passing it to `EachReducer`, but I don't think it's possible with the current `ReducerBuilder` (I didn't check yet). Or it could be a parameter of `ForEachReducer`, as this is currently the one that can see both `ParentState` and `EachState`. This is maybe the simplest approach. `LazyForEachStore` would be natural too. `ForEachStore` is currently held back by its deprecated forms, so having a new type would allow to make it work more efficiently.

About `IdentifiedCollection`, I still don't know if `protocol IdentifiedCollection` makes sense in `swift-indentified-collection`. This is a vast topic, and the `OrderedSet<ID>` is maybe too restrictive too. For example, what would you do with a collection similar to [this one](https://developer.apple.com/documentation/swiftui/indexedidentifiercollection) if it'd become more exposed? *(EDIT: [or these ones](https://github.com/apple/swift-collections/tree/feature/SortedCollections/Sources/SortedCollections))*

But in this library, I still think that extracting efficiently an `OrderedDictionary` from `IdentifiedArray` makes sense:
```swift
extension OrderedDictionary {
  public init(_ identifiedArray: IdentifiedArray<Key, Value>) {
    self = identifiedArray._dictionary
  }
}
```
Being an `IdentifiedArray` plays no different role than `OrderedDictionary` in `ForEachStore/Reducer`, so it would be natural to define these types based on `OrderedDictionary` and provide `IdentifiedArray` initializers that would downcast to `OrderedDictionary`. The `IdentifiedArray -> OrderedDictionary` transform seems harmless as long as `IdentifiedArray` is backed by `OrderedDictionary`, which seems to be the case for the foreseeable future.

**Reply by iampatbrown** [reply]
> About IdentifiedCollection, I still don't know if protocol IdentifiedCollection makes sense in swift-indentified-collection

Agreed. 

> I still think that extracting efficiently an OrderedDictionary from IdentifiedArray makes sense

Yeah, that sounds pretty good. 

> About DynamicViewContent, after having played a little with it, I don't think this is a thing for us to implement. 

Something to look into for sure. I haven't explored `DynamicViewContent` much.

**Reply by tgrapperon** [reply]
Ha, I didn't tried to implement it, but I just realized that extracting an `OrderedDictionary` won't play well with the setter part in `ForEachReducer`. So your initial parametric solution seems to be the solution with the least side effects for `ForEachReducer`. But `ForEachStore` could be implemented with `OrderedDictionary` if one can initialize them from `IdentifiedArray`.
Yesterday again I had a case where `OrderedDictionary` variants would have allowed me to avoid wrapping a fixed array of things into `IdentifiedArray<Int, Identified<Element>>` and directly work with a simpler, and yet relatively safe, `OrderedDictionary<Int, Element>`, so I can definitely feel the appeal!

**Comment by heyltsjay** [comment]
Any chance we can try to get this into 1.0? I feel confident that an ordered dictionary is a nice data structure to model state. Leaves shouldn't necessarily have to carry their own ID or be aware of their usage context.

**Reply by kaishin** [reply]
I second this. We use identified array most of the time but there are cases where dictionaries make more sense.
